### DISCUSSION TITLE
Drop support for outdated PHP versions

### DISCUSSION THREAD

**Discussion by Rumperuu** [markcheret/footnotes#167] [Ideas]
I've been going through #35 recently and I keep running into restrictions due to the fact that the plugin has to support PHP 7.0. For example, [nullable types and `void` as a function return type were not added](https://www.php.net/manual/en/migration71.new-features.php) until 7.1, and [class property typing](https://www.php.net/manual/en/language.types.declarations.php) didn't come in until 7.4.

I propose that we increase the minimum supported PHP version to 7.4, in line with the [current minimum for WordPress](https://wordpress.org/about/requirements/) itself. One the one hand, only [28.3% of WordPress installations](https://wordpress.org/about/stats/) have 7.4+, but on the other hand all PHP [versions <7.3](https://www.php.net/supported-versions.php) are officially EOL and should be considered insecure – 7.3 is active support EOL and will lose security support at the end of this year. 

What I am thinking is that the big 3.0 release can support 7.4+ (and peg its minimum version to WordPress' from then on), but anyone on an older version of PHP will still be able to use the last 2.x.x release if they absolutely insist. In the unlikely event that there are any security updates in the future we can backport them (but no features, bugfixes, etc). Plus, the stats for plugin version installations suggest that there are a fair few dead sites using our plugin, so they shouldn't be affected as they clearly aren't updating.

On a related, but separate, note, I am also looking into [Rector](https://github.com/rectorphp/rector/) as a way of enabling us to write code that takes advantage of PHP 8.x features whilst also transpiling down to PHP 7.x-compatible code.

So, in summary:

- 3.0 should require PHP 7.4+;
- after 3.0's release, 2.x.x will receive only security updates; and
- if Rector works out, then the codebase should adopt PHP 8.x as its development target and transpile down.

**Comment by markcheret** [comment]
Fully agree on all points!

**Comment by markcheret** [comment]
I suggest security fixes for 2.x only for critical vulnerabilities. For the rest it's time to update.

**Reply by Rumperuu** [reply]
On a related note, any idea where we've derived the ‘Requires at least’ WP version (3.9) from? And if it's still accurate? https://github.com/markcheret/footnotes/blob/4fcd349252941136e6b4a16dc13a83c47d83b97b/src/readme.txt#L5

**Reply by markcheret** [reply]
that is historic baggage. I'd bump it up considerably.
Also, I'd love a compatibility testing matrix, so it's "verified" that it's true what is claimed here :)
